JDK-8297408: Consolidate code in runtime/ErrorHandling

tstuefe opened the pull request:
There is a lot of duplicate coding in runtime/ErrorHandling that could be cut down using the new HsErrFileUtils utilities added with [JDK-8296906](https://bugs.openjdk.org/browse/JDK-8296906).

This patch mainly cuts away duplicate coding. Very minor functional changes:
- All tests will now use try-with-resources when opening hs-err files
- [TestCrashOnOutOfMemoryError.java](https://github.com/openjdk/jdk/pull/11283/files#diff-7560f2e2700932e015563716b3b0880495922c4c5731ddadba3b33bd937358bd) now checks for at least one matching line in the hs-err file - before, it basically just scanned for the END marker, which seemed pointless
- [test/hotspot/jtreg/runtime/ErrorHandling/TestDwarf.java](https://github.com/openjdk/jdk/pull/11283/files#diff-c04dd18d5a29a42e3a9a008c0b917663a7617c8c9a0d6e9efe8755773260ea95) : heaved Pattern construction out of loop
- [test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java](https://github.com/openjdk/jdk/pull/11283/files#diff-086956c42403a16043a111b0582736ac21ce0522dba866f5f9bc3fdb16037a0c) : I tried to preserve the debug output that was painstakingly added by @dcubed-ojdk apart from one change: I now printe the hs-err file as part of the scanning process. This works fine since in case of success, it will only be parsed as far as all patterns are matched, which happens after five-ish lines. Only if none are found, we scan and therefore print the whole hs-err file. So the old behavior is preserved in spirit.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297408](https://bugs.openjdk.org/browse/JDK-8297408): Consolidate code in runtime/ErrorHandling


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [b9be700a](https://git.openjdk.org/jdk/pull/11283/files/b9be700a3f4c00feada2f5c0418cd1f7ca1f7dd5)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [85996592](https://git.openjdk.org/jdk/pull/11283/files/85996592af82e035d29cfc71c29237caee5cdeac)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11283/head:pull/11283` \
`$ git checkout pull/11283`

Update a local copy of the PR: \
`$ git checkout pull/11283` \
`$ git pull https://git.openjdk.org/jdk pull/11283/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11283`

View PR using the GUI difftool: \
`$ git pr show -t 11283`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11283.diff">https://git.openjdk.org/jdk/pull/11283.diff</a>

</details>

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
-     * Given an open hs-err file, read it line by line and check for pattern. Pattern
-     * need to appear in order, but not necessarily uninterrupted.
+     * Given an open hs-err file, read it line by line and check for existence of a pattern. Will fail
+     * if pattern are missing, or if the END marker is missing.
```
"patterns are".

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
     /**
-     * Given an open hs-err file, read it line by line and check for pattern. Pattern
-     * need to appear in order, but not necessarily uninterrupted.
+     * Given an open hs-err file, read it line by line and check for existence of a pattern. Will fail
```
"a set of patterns".

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
+     * Given an open hs-err file, read it line by line and check for various conditions.
+     * @param f input file
+     * @param positivePatterns Optional array of patterns that need to appear, in given order, in the file. Missing
+     *                        pattern cause the test to fail.
```
patterns

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
+     *                        pattern cause the test to fail.
+     * @param negativePatterns Optional array of patterns that must not appear in the file; test fails if they do.
+     *                        Order is irrelevant.
+     * @param checkEndMarker If true, we check for the final "END" in an hs-err file; it missing indicates hs-err file
```
"if it is missing it indicates"

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
-                System.out.println(line);
+    /**
+     * Given an open hs-err file, read it line by line and check for various conditions.
+     * @param f input file
+     * @param positivePatterns Optional array of patterns that need to appear, in given order, in the file. Missing
```
Consistency check: Sometimes you start the description with a capital letter and other times not. I don't know what preferred JDK Java style is.

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
-                    currentPattern ++;
                 }
+                lastLine = line;
+                lineNo ++;
```
No space before unary operators.

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/TestDwarf.java:
```diff
@@ -123,18 +123,15 @@ private static void runAndCheck(Flags flags, DwarfConstraint... constraints) thr
-            BufferedReader reader = new BufferedReader(new FileReader(hs_err_file));
+
+        File hs_err_file = HsErrFileUtils.openHsErrFileFromOutput(crashOut);
+        try (BufferedReader reader = new BufferedReader(new FileReader(hs_err_file))) {
```
Not sure this is the correct twr idiom when constructing multiple resources ie should it be:
```
try (var fr = new FileReader(hs_err_file);
      var reader = new BufferredReader(fr);) {
...
}
```

dholmes-ora commented on test/hotspot/jtreg/runtime/ErrorHandling/BadNativeStackInErrorHandlingTest.java:
```diff
@@ -58,44 +58,15 @@ public static void main(String[] args) throws Exception {
-    } else {
-        System.out.println("End marker found.");
-    }
+    HsErrFileUtils.checkHsErrFileContent(hs_err_file, null, negativePatterns, true, false);
```
Can we annotate the true/false parameters so the reader doesn't have to go and look up what they mean i.e. `true /* check END */, false /*not verbose */` ?

tstuefe replied:
My IDE usually does this for me, but I can make it explicit, sure.

dholmes-ora reviewed (commented):
Overall this seems quite reasonable. A few minor nits below.

Thanks.

tstuefe commented:
David, does this pass for you?

Any other review?

MBaesken reviewed (approved):
Looks good to me, minor points 
test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java

"existence of a set of pattern"  shouldn't this be
"existence of a set of patterns"  ?

please check COPYRIGHT headers , e.g.
test/hotspot/jtreg/runtime/ErrorHandling/NestedThreadsListHandleInErrorHandlingTest.java

tstuefe commented:
Fixed comments as @MBaesken suggested; fixed copyrights. I took the freedom of adding SAP copyrights - with the correct introduction data - for all tests I personally introduces between 2013 and 2019. I had forgotten to do that.

tstuefe commented:
@dholmes-ora are we good here?

dholmes-ora reviewed (approved):
Sorry for the delay.

One minor issue with a copyright line in one file that needs fixing, but otherwise this seems okay to me.

Thanks.

dholmes-ora commented:
Running through our CI just to be cautious.

dholmes-ora commented:
Our CI passed tiers 1-3

tstuefe commented:
Thank you David!

I re-ran 32-bit tests on x86 and arm too.

/integrate
